File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.cpp
class Solution {
public:
    long long countSubarrays(vector<int>& nums, int minK, int maxK) {
        long long ans = 0;
        int lastMinIdx = -1, lastMaxIdx = -1, subArrStartIdx = 0;
        
        for(int i = 0; i < size(nums); ++i) {
            // invalid element, hence start fresh with new starting index of subarrays
            if(nums[i] < minK or nums[i] > maxK) {
                lastMinIdx = lastMaxIdx = -1;
                subArrStartIdx = i + 1;
            }
            // update the indices where we get our minK & maxK
            if(nums[i] == minK) lastMinIdx = i;
            if(nums[i] == maxK) lastMaxIdx = i;
            
            // the subarray ends at nums[i]
            // and starts in the range: [subArrStartIdx, min(lastMinIdx, lastMaxIdx)]
            // thus, we have min(lastMinIdx, lastMaxIdx) - subArrStartIdx + 1 choices 
            // until we find an invalid element
            ans += max(0, min(lastMinIdx, lastMaxIdx) - subArrStartIdx + 1);
        }
        return ans;
    }
};

File: 2444-count-subarrays-with-fixed-bounds/2444-count-subarrays-with-fixed-bounds.py
class Solution:
    def countSubarrays(self, nums: List[int], minK: int, maxK: int) -> int:
        count, lastMinIdx, lastMaxIdx, subarrStartIdx = 0, -1, -1, 0
        for i in range (0, len(nums)):
            if nums[i] < minK or nums[i] > maxK:
                lastMinIdx, lastMaxIdx, subarrStartIdx = -1, -1, i + 1
            if nums[i] == minK:
                lastMinIdx = i
            if nums[i] == maxK:
                lastMaxIdx = i
            count += max(0, min(lastMinIdx, lastMaxIdx) - subarrStartIdx + 1)
        return count